architecture/2025-03-upgrade-design-doc.md

```diff
@@ -0,0 +1,677 @@
+    UserDataRestore --> RollbackHelm[Rollback Helm changes]
+    RollbackHelm --> ReleaseLock
+    ReleaseLock --> End[Display results to user]
+```
+
+**Version Detection and Validation:**
+
+The upgrade process begins with detecting the currently installed Radius version (the versions of the components in the cluster) and validating it against the requested target version. We can build an interface (or use/improve an existing one) to achieve this:
+
+```go
+// VersionValidator interface
+type VersionValidator interface {
+    ValidateTargetVersion(currentVersion, targetVersion string) error
+    GetLatestVersion() (string, error)
+    IsValidVersion(version string) bool
+}
+```
+
+This interface will be implemented (or existing will be improved) to handle version comparisons, prevent downgrades, and resolve the "latest" version tag to a specific version number. The implementation will (probably) connect to the GitHub API to fetch available release versions when needed.
+
+**Upgrade Lock Mechanism:**
+
+To prevent concurrent modifications during upgrades, we'll implement a lock system that's acquired early in the process:
+
+```go
+type UpgradeLock interface {
+    // Acquires the upgrade lock, preventing other commands from modifying data
+    AcquireLock(ctx context.Context) error
+
+    // Releases the upgrade lock, allowing other commands to proceed
+    ReleaseLock(ctx context.Context) error
+
+    // Checks if an upgrade is currently in progress
+    IsUpgradeInProgress(ctx context.Context) (bool, error)
+}
+```
+
+We can utilize Kubernetes Lease objects (coordination.k8s.io/v1) for implementing the distributed locking mechanism (open to discussion and suggestions). Leases are purpose-built for this use case, providing built-in lease duration and automatic expiration capabilities. For more information, see: <https://kubernetes.io/docs/concepts/architecture/leases/>.
+
+Other CLI commands (`rad deploy app.bicep`, `rad delete app my-app` or other data-changing commands) that modify data will check for this lock before proceeding:
+
+```go
+inProgress, _ := upgradeLock.IsUpgradeInProgress(ctx)
+if inProgress {
+    return errors.New("An upgrade is currently in progress. Please try again after the upgrade completes.")
+}
+```
+
+**Pre-flight Check System:**
+
+Pre-flight checks run before any changes are made to ensure the upgrade can proceed safely.
+
+```go
+type PreflightCheck interface {
+    Run(ctx context.Context) (bool, string, error)
+    Name() string
+    // Error, Warning, Info
+    Severity() CheckSeverity
+}
+```
+
+Checks will include:
+
+1. Version compatibility verification
+2. Existing installation detection
+3. Database connectivity
+4. Custom configuration validation
+
+**User Data Backup and Restore System:**
+
+Rather than taking complete snapshots of the underlying databases (etcd/PostgreSQL), we'll implement a more targeted approach that backs up only the user application metadata and configuration that Radius manages:
```

Comment:
What other data won't be backed up?

Reply:
Everything other than Radius user related data.